From 2c53f37051832d9adf4eb9ca539a5dba4c108de4 Mon Sep 17 00:00:00 2001
From: Christophe Roullier <christophe.roullier@st.com>
Date: Thu, 6 Apr 2023 11:35:59 +0200
Subject: [PATCH 484/529] net: stmmac: fix pinctrl management during
 suspend/resume

In the deepest low power modes the pinctrl are lost and never are restored if
the interface is down.
This commit makes sure that pinctrl modes are set in all the cases.

Signed-off-by: Arturo Buzarra <arturo.buzarra@digi.com>
Signed-off-by: Christophe Roullier <christophe.roullier@st.com>
Change-Id: I6f9f5eb8aa31f7321812f4ec38c9f6134e744b97
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/299336
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Alexandre TORGUE <alexandre.torgue@foss.st.com>
Domain-Review: Alexandre TORGUE <alexandre.torgue@foss.st.com>
---
 drivers/net/ethernet/stmicro/stmmac/stmmac_main.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -7456,8 +7456,11 @@ int stmmac_suspend(struct device *dev)
 	struct stmmac_priv *priv = netdev_priv(ndev);
 	u32 chan;
 
-	if (!ndev || !netif_running(ndev))
+	if (!ndev || !netif_running(ndev)) {
+		/* Select sleep pin state */
+		pinctrl_pm_select_sleep_state(dev);
 		return 0;
+	}
 
 	mutex_lock(&priv->lock);
 
@@ -7564,8 +7567,11 @@ int stmmac_resume(struct device *dev)
 	struct stmmac_priv *priv = netdev_priv(ndev);
 	int ret;
 
-	if (!netif_running(ndev))
+	if (!netif_running(ndev)) {
+		/* Select default pin state */
+		pinctrl_pm_select_default_state(priv->device);
 		return 0;
+	}
 
 	/* Power Down bit, into the PM register, is cleared
 	 * automatically as soon as a magic packet or a Wake-up frame
